From 1f22100cbb9d688bab9c45bfa6f19a67ff8ef994 Mon Sep 17 00:00:00 2001
From: Kishan Mochi <kishan.mochi@intel.com>
Date: Tue, 8 Mar 2022 13:06:23 +0530
Subject: [PATCH] Port SRIOV graphics and media installation for host

Changes:
- added function to install components for SRIOV
- ported SRIOV installation script
- improved grub update logic
- removed old mesa installation patch

diff --git a/scripts/setup_host.sh b/scripts/setup_host.sh
index f6043e5..04ae987 100755
--- a/scripts/setup_host.sh
+++ b/scripts/setup_host.sh
@@ -34,6 +34,43 @@ function ubu_changes_require(){
     sudo apt install -y libglib2.0-dev libncurses-dev libuuid1 uuid-dev libjson-c-dev
 }
 
+function ubu_install_sriov(){
+    # Switch to Xorg for Ubuntu 21.04
+    if [[ $(lsb_release -rs) == "21.04" ]]; then
+        sed -i "s/\#WaylandEnable=false/WaylandEnable=false/g" /etc/gdm3/custom.conf
+    fi
+
+    sudo apt install -y net-tools openssh-server git make autoconf libtool meson vim v4l-utils pkg-config libpciaccess-dev cmake \
+                        python3-pip python3.10 llvm-12-dev libelf-dev bison flex wayland-protocols libwayland-dev libwayland-egl-backend-dev \
+                        libx11-dev libxext-dev libxdamage-dev libx11-xcb-dev libxcb-glx0-dev libxcb-dri2-0-dev libxcb-dri3-dev \
+                        libxcb-present-dev libxshmfence-dev libxxf86vm-dev libxrandr-dev libkmod-dev libprocps-dev libdw-dev libpixman-1-dev \
+                        libcairo2-dev libudev-dev libgudev-1.0-0 gtk-doc-tools sshfs mesa-utils xutils-dev libunwind-dev libxml2-dev doxygen \
+                        xmlto cmake libpciaccess-dev graphviz libjpeg-dev libwebp-dev libsystemd-dev libdbus-glib-1-dev libpam0g-dev \
+                        freerdp2-dev libxkbcommon-dev libinput-dev libxcb-shm0-dev libxcb-xv0-dev libxcb-keysyms1-dev libxcb-randr0-dev \
+                        libxcb-composite0-dev libxcursor-dev liblcms2-dev libpango1.0-dev libglfw3-dev libxcb-composite0-dev libxcursor-dev \
+                        libgtk-3-dev libsdl2-dev virtinst virt-viewer virt-manager libspice-server-dev libusb-dev libxfont-dev libxkbfile-dev \
+                        libepoxy-dev rpm libncurses5-dev libncursesw5-dev liblz4-tool git-lfs uuid mtools python3-usb python3-pyudev \
+                        libjson-c-dev libfdt-dev socat bridge-utils uml-utilities python-dev libcap-ng-dev libusb-1.0-0-dev nasm acpidump \
+                        iasl libseccomp-dev libtasn1-6-dev libgnutls28-dev expect gawk
+    sudo apt install -y python3-mako
+
+    # Clean up any existing folders
+    function del_existing_folder() {
+        if [ -d "$1" ]; then
+            echo "Deleting existing folder $1"
+            rm -fr $1
+        fi
+    }
+    del_existing_folder $CIV_WORK_DIR/media
+    del_existing_folder $CIV_WORK_DIR/gstreamer
+    del_existing_folder $CIV_WORK_DIR/graphics
+    del_existing_folder $CIV_WORK_DIR/neo
+
+    # Start setup
+    source $CIV_WORK_DIR/scripts/sriov_setup_host.sh
+
+}
+
 function ubu_install_qemu_gvt(){
     sudo apt purge -y "^qemu"
     sudo apt autoremove -y
@@ -152,12 +189,15 @@ function install_vm_manager() {
 }
 
 function ubu_enable_host_gvt(){
-    if [[ ! `cat /etc/default/grub` =~ "i915.enable_gvt=1 intel_iommu=on i915.force_probe=*" ]]; then
+    if [[ ! `cat /etc/default/grub` =~ "i915.enable_gvt=1" ]]; then
         read -p "The grub entry in '/etc/default/grub' will be updated for enabling GVT-g and GVT-d, do you want to continue? [Y/n]" res
         if [ x$res = xn ]; then
-            exit 0
+            return
+        fi
+        if [[ ! `cat /etc/default/grub` =~ "intel_iommu=on i915.force_probe=*" ]]; then
+            sed -i "s/GRUB_CMDLINE_LINUX=\"/GRUB_CMDLINE_LINUX=\"intel_iommu=on i915.force_probe=* /g" /etc/default/grub
         fi
-        sed -i "s/GRUB_CMDLINE_LINUX=\"/GRUB_CMDLINE_LINUX=\"i915.enable_gvt=1 intel_iommu=on i915.force_probe=*/g" /etc/default/grub
+        sed -i "s/GRUB_CMDLINE_LINUX=\"/GRUB_CMDLINE_LINUX=\"i915.enable_gvt=1 /g" /etc/default/grub
         update-grub
 
         echo -e "\nkvmgt\nvfio-iommu-type1\nvfio-mdev\n" >> /etc/initramfs-tools/modules
@@ -168,24 +208,21 @@ function ubu_enable_host_gvt(){
 }
 
 function ubu_enable_host_sriov(){
-   if [[ ! `cat /etc/default/grub` =~ "i915.enable_guc=0x7" ]]; then
+    if [[ ! `cat /etc/default/grub` =~ "i915.enable_guc=0x7 udmabuf.list_limit=8192" ]]; then
         read -p "The grub entry in '/etc/default/grub' will be updated for enabling SRIOV, do you want to continue?     [Y/n]" res
         if [ x$res = xn ]; then
-            exit 0
+            return
+        fi
+        if [[ ! `cat /etc/default/grub` =~ "intel_iommu=on i915.force_probe=*" ]]; then
+            sed -i "s/GRUB_CMDLINE_LINUX=\"/GRUB_CMDLINE_LINUX=\"intel_iommu=on i915.force_probe=* /g" /etc/default/grub
         fi
-	sed -i "s/GRUB_CMDLINE_LINUX=\"/GRUB_CMDLINE_LINUX=\"i915.enable_guc=0x7 udmabuf.list_limit=8192  /g" /etc/default/grub
+        sed -i "s/GRUB_CMDLINE_LINUX=\"/GRUB_CMDLINE_LINUX=\"i915.enable_guc=0x7 udmabuf.list_limit=8192 /g" /etc/default/grub
         update-grub
 
         echo -e "\nkvmgt\nvfio-iommu-type1\nvfio-mdev\n" >> /etc/initramfs-tools/modules
         update-initramfs -u -k all
 
-        # Switch to Xorg for Ubuntu 21.04
-        if [[ $(lsb_release -rs) == "21.04" ]]; then
-            sed -i "s/\#WaylandEnable=false/WaylandEnable=false/g" /etc/gdm3/custom.conf
-        fi
-
         reboot_required=1
-
     fi
 }
 
@@ -494,6 +531,7 @@ check_network
 check_kernel_version
 
 ubu_changes_require
+ubu_install_sriov
 ubu_install_qemu_gvt
 ubu_build_ovmf_gvt
 ubu_enable_host_gvt
diff --git a/scripts/sriov_setup_host.sh b/scripts/sriov_setup_host.sh
new file mode 100644
index 0000000..8be0eaa
--- /dev/null
+++ b/scripts/sriov_setup_host.sh
@@ -0,0 +1,214 @@
+#!/bin/bash
+RED='\033[0;31m'
+NC='\033[0m'
+GREEN='\033[0;32m'
+YELLOW='\033[1;33m'
+
+export PrefixPath=/usr
+export LibPath=/usr/lib/x86_64-linux-gnu
+export nproc=20
+export WrkDir=`pwd`
+
+function check_build_error(){
+        if [ $? -ne 0 ]; then
+                echo -e "${RED}$1: Build Error ${NC}"
+                exit -1
+        else
+                echo -e "${GREEN}$1: Build Success${NC}"
+        fi
+}
+
+git config --global advice.detachedHead false
+#media
+echo "export LIBVA_DRIVER_NAME=iHD" | sudo tee -a /etc/environment
+echo "export LIBVA_DRIVERS_PATH=/usr/lib/x86_64-linux-gnu/dri" | sudo tee -a /etc/environment
+echo "export GIT_SSL_NO_VERIFY=true" | sudo tee -a /etc/environment
+source /etc/environment
+
+git lfs install --skip-smudge
+git clone https://gitlab.freedesktop.org/mesa/drm.git media/libdrm
+cd media/libdrm
+git checkout refs/tags/libdrm-2.4.107
+meson build/ --prefix=$PrefixPath --libdir=$LibPath
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/intel/libva.git media/libva
+cd media/libva
+git checkout refs/tags/2.13.0
+meson build/ --prefix=$PrefixPath --libdir=$LibPath
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/intel/libva-utils.git media/libva-utils
+cd media/libva-utils
+git checkout refs/tags/2.13.0
+meson build/ --prefix=$PrefixPath --libdir=$LibPath
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/intel/gmmlib.git media/gmmlib
+cd media/gmmlib
+git checkout refs/tags/intel-gmmlib-21.3.3
+mkdir build && cd build
+cmake -DCMAKE_INSTALL_PREFIX=$PrefixPath ../
+make -j "$(nproc)"
+check_build_error
+sudo make install
+cd $WrkDir
+
+
+git clone https://github.com/intel/media-driver.git media/media-driver
+cd media/media-driver
+git checkout refs/tags/intel-media-21.3.5
+git apply $CIV_WORK_DIR/vertical_patches/host/media/media-driver/*.patch
+
+mkdir build_media && cd build_media
+cmake ../ -DCMAKE_INSTALL_PREFIX=$PrefixPath
+make -j "$(nproc)"
+check_build_error
+sudo make install
+cd $WrkDir
+# Create igfx_user_feature.txt
+echo "[KEY]"                                     | sudo tee -a /etc/igfx_user_feature.txt
+echo "    0x00000001"                            | sudo tee -a /etc/igfx_user_feature.txt
+echo "    UFKEY_INTERNAL\LibVa"                  | sudo tee -a /etc/igfx_user_feature.txt
+echo "        [VALUE]"                           | sudo tee -a /etc/igfx_user_feature.txt
+echo "            Disable MMC"                   | sudo tee -a /etc/igfx_user_feature.txt
+echo "            4"                             | sudo tee -a /etc/igfx_user_feature.txt
+echo "            1"                             | sudo tee -a /etc/igfx_user_feature.txt
+echo "        [VALUE]"                           | sudo tee -a /etc/igfx_user_feature.txt
+echo "            Enable HCP Scalability Decode" | sudo tee -a /etc/igfx_user_feature.txt
+echo "            4"                             | sudo tee -a /etc/igfx_user_feature.txt
+echo "            0"                             | sudo tee -a /etc/igfx_user_feature.txt
+echo "[KEY]"                                     | sudo tee -a /etc/igfx_user_feature.txt
+echo "    0x00000002"                            | sudo tee -a /etc/igfx_user_feature.txt
+echo "    UFKEY_INTERNAL\Report"                 | sudo tee -a /etc/igfx_user_feature.txt
+
+
+#onevpl-gpu
+git clone https://github.com/oneapi-src/oneVPL-intel-gpu.git media/oneVPL-gpu
+cd media/oneVPL-gpu
+git checkout refs/tags/intel-onevpl-21.3.4
+git apply $CIV_WORK_DIR/vertical_patches/host/media/oneVPL-gpu/*.patch
+
+mkdir build && cd build
+cmake ../ -DCMAKE_INSTALL_PREFIX=$PrefixPath
+make -j "$(nproc)"
+check_build_error
+sudo make install
+cd $WrkDir
+
+#onevpl
+git clone https://github.com/oneapi-src/oneVPL.git media/oneVPL
+cd media/oneVPL
+git checkout refs/tags/v2021.6.0
+git apply $CIV_WORK_DIR/vertical_patches/host/media/oneVPL/*.patch
+mkdir build && cd build
+cmake ../ -DCMAKE_INSTALL_PREFIX=$PrefixPath
+make -j "$(nproc)"
+check_build_error
+sudo make install
+cd $WrkDir
+
+
+#gstreamer
+git clone https://github.com/GStreamer/gstreamer.git gstreamer/gstreamer
+cd gstreamer/gstreamer
+git checkout b4ca58df7624b005a33e182a511904d7cceea890
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/GStreamer/gst-plugins-base.git gstreamer/gst-plugins-base
+cd gstreamer/gst-plugins-base
+git checkout ce937bcb21412d7b3539a2da0509cc96260562f8
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/GStreamer/gst-plugins-good.git gstreamer/gst-plugins-good
+cd gstreamer/gst-plugins-good
+git checkout 20bbeb5e37666c53c254c7b08470ad8a00d97630
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/GStreamer/gst-plugins-bad.git gstreamer/gst-plugins-bad
+cd gstreamer/gst-plugins-bad
+git checkout ca8068c6d793d7aaa6f2e2cc6324fdedfe2f33fa
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/GStreamer/gst-plugins-ugly.git gstreamer/gst-plugins-ugly
+cd gstreamer/gst-plugins-ugly
+git checkout 499d3cd726a4ca9cbbdd4b4fe9ccdca78ef538ba
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/GStreamer/gstreamer-vaapi.git gstreamer/gstreamer-vaapi
+cd gstreamer/gstreamer-vaapi
+git checkout c3ddb29cb2860374f9efbed495af7b0eead08312
+git apply $CIV_WORK_DIR/vertical_patches/host/gstreamer/gstreamer-vaapi/*.patch
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+git clone https://github.com/GStreamer/gst-rtsp-server.git gstreamer/gst-rtsp-server
+cd gstreamer/gst-rtsp-server
+git checkout 0b037e35e7ed3259ca05be748c382bc40e2cdd91
+meson build --prefix=$PrefixPath --libdir=$LibPath -Dgtk_doc=disabled
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+
+
+#mesa
+git clone https://gitlab.freedesktop.org/mesa/mesa.git graphics/mesa
+cd graphics/mesa
+git checkout 0e0633ca49425dbc869521cede6a82d2d91c8042
+git apply $CIV_WORK_DIR/vertical_patches/host/graphics/mesa/*.patch
+meson build/ --prefix=$PrefixPath -Dgallium-drivers="swrast,iris,kmsro" -Dvulkan-drivers=intel -Ddri-drivers=i965
+ninja -C build && sudo ninja -C build install
+check_build_error
+cd $WrkDir
+# Create mesa_driver.sh
+echo "is_vf=\`dmesg | grep \"SR-IOV VF\"\`"         | sudo tee -a /etc/profile.d/mesa_driver.sh
+echo "if [[ \$is_vf =~ \"VF\" ]]; then"                | sudo tee -a /etc/profile.d/mesa_driver.sh
+echo "    export MESA_LOADER_DRIVER_OVERRIDE=pl111" | sudo tee -a /etc/profile.d/mesa_driver.sh
+echo "else"                                         | sudo tee -a /etc/profile.d/mesa_driver.sh
+echo "    export MESA_LOADER_DRIVER_OVERRIDE=iris"  | sudo tee -a /etc/profile.d/mesa_driver.sh
+echo "fi"                                           | sudo tee -a /etc/profile.d/mesa_driver.sh
+
+
+#OpenCL
+mkdir neo
+cd neo
+wget https://github.com/intel/compute-runtime/releases/download/21.47.21710/intel-gmmlib-devel_21.3.3_amd64.deb
+wget https://github.com/intel/compute-runtime/releases/download/21.47.21710/intel-gmmlib_21.3.3_amd64.deb
+wget https://github.com/intel/intel-graphics-compiler/releases/download/igc-1.0.9389/intel-igc-core_1.0.9389_amd64.deb
+wget https://github.com/intel/intel-graphics-compiler/releases/download/igc-1.0.9389/intel-igc-opencl_1.0.9389_amd64.deb
+wget https://github.com/intel/compute-runtime/releases/download/21.47.21710/intel-opencl-icd_21.47.21710_amd64.deb
+wget https://github.com/intel/compute-runtime/releases/download/21.47.21710/intel-level-zero-gpu_1.2.21710_amd64.deb
+sudo dpkg -i *.deb
+cd $WrkDir
-- 
2.35.1

